src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.cs

```diff
@@ -165,12 +152,24 @@ internal AppleCertificatePal(SafeSecCertificateHandle certHandle)
             _certHandle = certHandle;
         }
 
-        internal AppleCertificatePal(SafeSecIdentityHandle identityHandle)
+        internal AppleCertificatePal(SafeSecIdentityHandle identityHandle, SafeTemporaryKeychainHandle? temporaryKeychainHandle)
         {
             Debug.Assert(!identityHandle.IsInvalid);
 
             _identityHandle = identityHandle;
             _certHandle = Interop.AppleCrypto.X509GetCertFromIdentity(identityHandle);
+
+            if (temporaryKeychainHandle != null)
+            {
+                // If we used temporary keychain we need to prevent deletion.
+                // on 10.15+ if keychain is unlinked, certain certificate operations may fail.
+                bool success = false;
+                temporaryKeychainHandle.DangerousAddRef(ref success);
```

Comment:
I no longer see a balancing DangreousRelease, meaning the temporary keychains that are created during a test are probably living on disk forever now.

Reply:
It's already present in the AppleCertificatePal.Dispose method. I just reused the existing mechanism that was added there to workaround earlier issue.

Reply:
https://github.com/dotnet/runtime/blob/794a0a84ae97b3b73425abf023250a391e8197e0/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.cs#L183-L187

Reply:
On second thought I didn't check whether the Dispose method is always called though.

Reply:
There's also a difference between SafeHandle.Dispose and SafeHandle.DangerousRelease: Dispose only runs once, DangerousRelease has an effect on every call.  (Dispose undoes the single refcount from allocation, DangerousRelease undoes the refcount modification from DangerousAddRef)

So if this DangerousAddRef gets hit, the object should never hit a 0 count, so it'll never call ReleaseHandle.

Reply:
That all makes sense. The `DangerousAddRef` has a paired `DangerousRelease`. The PR takes an existing code path with `_tempKeychain` and adds the same logic to few more places.

The problem is that the `AppleCertificatePal.Dispose` is likely not guaranteed to be called (ie. no finalizer on `AppleCertificatePal` and/or `X509Certificate`) which is pre-existing problem.

Reply:
> The problem is that the AppleCertificatePal.Dispose is likely not guaranteed to be called (ie. no finalizer on AppleCertificatePal and/or X509Certificate) which is pre-existing problem.

Well, it's not normally a problem.  The SafeHandle itself is finalizable, so if the AppleCertificatePal gets garbage collected then the keychain handle can get GCd and then the finalizer runs, drops the refcount again, and everything's happy.

But it looks like I used a Dictionary instead of a ConditionalWeakTable in the current tracker, so there's always a strong reference, so the current code never finalizes.  Whoops.

Reply:
Oh yeah, I didn't even notice the strong references in the current tracker. What I was worried about is this:
https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.cs#L115-L125
https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/AppleCertificatePal.cs#L184-L188
As far as I can see (just quick check from phone) there are no finalizer here, so this `Dispose` may not get called and leak.